### PR TITLE
feat(TPSVC-15467): show list items number

### PR DESCRIPTION
**Pull request by lmaillet**
**What is the problem this PR is trying to solve?**
The total number of items can be shown in the list.
See https://jira.talendforge.org/browse/TPSVC-15467

**What is the chosen solution to this problem?**

For **List**
- [Show `x items` in the right of the list toolbar](http://2935.talend.surge.sh/components/?path=/story/data-list-list--table-with-number-of-items)
- [Show `y/x items` when items are selected and remove the selected number in the left](http://2935.talend.surge.sh/components/?path=/story/data-list-list--selection-with-number-of-items)
- Be able to override the label `items` with another title
- [No regression on the current selection behavior](http://2935.talend.surge.sh/components/?path=/story/data-list-list--selection)

And the same for **ListComposition**
- [Show `x items` in the right of the list toolbar](http://2935.talend.surge.sh/components/?path=/story/data-list-list-composition--total-items)
- [Show `y/x items` when items are selected and remove the selected number in the left](http://2935.talend.surge.sh/components/?path=/story/data-list-list-composition--selectable-items-total-items), the number of selected items must be hidden in the ActionBar with the prop `hideCount`
- Be able to override the label `items` with another title
- [No regression on the current selection behavior](http://2935.talend.surge.sh/components/?path=/story/data-list-list-composition--selectable-items-actionbar)

**Please check if the PR fulfills these requirements**

* [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
* [x] Tests for the changes have been added (for bug fixes / features) And [non reg](./screenshots.md) done before need review
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/components/src/List/Toolbar/Toolbar.component.js [hunk 1]
```diff
@@ -84,9 +86,15 @@ function Toolbar({
 		<div className={classNames(theme['tc-list-toolbar'], 'tc-list-toolbar')}>
 			{injected('before-actionbar')}
 			{actionBar && (
-				<Renderer.ActionBar {...actionBarProps} className="list-action-bar">
+				<Renderer.ActionBar {...actionBarProps} hideSelected={!!itemsNumber} className="list-action-bar">
```
Comment (lmaillet) [inline packages/components/src/List/Toolbar/Toolbar.component.js:89]: Hide the existing selected label in the left of the `ToolBar` if `itemsNumber` is provided and will be shown in the right.

File: packages/components/src/List/Toolbar/ItemsNumber/ItemsNumber.component.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+						defaultValue: '{{selected}}/{{totalItems}} {{itemsLabel}}',
+						selected,
+						totalItems,
+						itemsLabel,
+				  })
+				: t('TOTAL_NUMBER_OF_ITEMS', {
+						defaultValue: '{{totalItems}} {{itemsLabel}}',
+						totalItems,
+						itemsLabel,
+				  })}
```
Comment (jsomsanith-tlnd) [inline packages/components/src/List/Toolbar/ItemsNumber/ItemsNumber.component.js:20]:
```suggestion
				? t('LIST_TOOLBAR_NUMBER_OF_SELECTED_ITEMS', {
						defaultValue: '{{count}}/{{total}} item',
						count: selected,
						total: totalItems,
				  })
				: t('LIST_TOOLBAR_TOTAL_NUMBER_OF_ITEMS', {
						defaultValue: '{{count}} item',
						count: totalItems,
				  })}

File: packages/components/src/List/Toolbar/ItemsNumber/ItemsNumber.component.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import getDefaultT from '../../../translate';
+
+function ItemsNumber({ id, totalItems, selected, label, t }) {
+	const itemsLabel = label || t('TOTAL_NUMBER_OF_ITEMS_LABEL', { defaultValue: 'items' });
```
Comment (jsomsanith-tlnd) [inline packages/components/src/List/Toolbar/ItemsNumber/ItemsNumber.component.js:6]:
```suggestion
```

File: packages/components/src/List/Toolbar/ItemsNumber/ItemsNumber.test.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	id: 'list-items-number',
+	totalItems: 42,
+};
+describe('ItemsNumber', () => {
+	it('should render total number of items', () => {
+		// when
+		const wrapper = shallow(<ItemsNumber {...props} />);
+
+		// then
+		expect(toJson(wrapper)).toMatchSnapshot();
```
Comment (jsomsanith-tlnd) [inline packages/components/src/List/Toolbar/ItemsNumber/ItemsNumber.test.js:16]: Please don't use snapshots for that. Snapshots should be used only to check the overall markup. You want to check the text, select it via enzyme and assert the expected value via jest please.

File: packages/components/src/List/ListComposition/ItemsNumber/ItemsNumber.component.test.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+		it('should render total number of items', () => {
+			// when
+			const wrapper = mount(
+				<ListContext.Provider value={defaultContext}>
+					<ItemsNumber {...props} />
+				</ListContext.Provider>,
+			);
+
+			// then
+			expect(wrapper.html()).toMatchSnapshot();
```
Comment (jsomsanith-tlnd) [inline packages/components/src/List/ListComposition/ItemsNumber/ItemsNumber.component.test.js:30]: Same here

File: packages/components/src/List/Toolbar/Toolbar.component.js [hunk 6]
```diff
@@ -84,9 +86,24 @@ function Toolbar({
 		<div className={classNames(theme['tc-list-toolbar'], 'tc-list-toolbar')}>
 			{injected('before-actionbar')}
 			{actionBar && (
-				<Renderer.ActionBar {...actionBarProps} className="list-action-bar">
+				<Renderer.ActionBar
+					{...actionBarProps}
+					hideSelected={!!itemsNumber}
```
[the remark this answers is not in the document]
Comment (lmaillet) [reply packages/components/src/List/Toolbar/Toolbar.component.js:91]: It's useless, `hideCount` comes from `ActionBar` props.

File: packages/components/src/List/ListComposition/ItemsNumber/ItemsNumber.component.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+						defaultValue: '{{selected}}/{{totalItems}} {{itemsLabel}}',
+						selected,
+						totalItems,
+						itemsLabel,
+				  })
+				: t('TOTAL_NUMBER_OF_ITEMS', {
+						defaultValue: '{{totalItems}} {{itemsLabel}}',
+						totalItems,
+						itemsLabel,
+				  })}
```
[the remark this answers is not in the document]
Comment (lmaillet) [reply packages/components/src/List/ListComposition/ItemsNumber/ItemsNumber.component.js:22]: `itemsLabel` can be provided as a prop to customize the view. For example `50 users`, `10/20 groups`.
So not sure it's possible to manage plural.

[the remark this answers is not in the document]
Comment (jsomsanith-tlnd) [reply packages/components/src/List/ListComposition/ItemsNumber/ItemsNumber.component.js]: You're right, to solve that, we can turn the label into a function `((count) => translated label)`. So if you want to customize it, you translate it in the app, and you can use the i18next-scanner to have the singular and plural in your app.
And by default, it's item/items as in the code currently

**Review by sgangnant — APPROVED** [review]

**Pull request merged**